### DISCUSSION TITLE
High (200+) XFRM interface count performance problem (throughput)

### DISCUSSION THREAD

**Discussion by karelvins** [strongswan/strongswan#216] [Q&A] [answered]
Hello,

I would like to ask you for help or maybe direction where to look for help.

I'm testing setup with higher number of XRFM interfaces and I'm facing throughput degradation with a growing number of created XRFM interfaces - not concurrent tunnels established but only XRFM interfaces precreated - even in DOWN state.
Issue is only unidirectional - from "client" to "hub". Throughput for traffic from hub to client is not affected.

I'm testing with iperf3 with 1 client connected - from client to hub:
2 interfaces - 1.36 Gbps
100 interfaces - 1.35 Gbps
200 interfaces - 1.19 Gbps
300 interfaces - 0.98 Gbps
500 interfaces - 0.71 Gbps

Throughput from hub to client is around 1.4 Gbps in all cases.

1 CPU core is 100%

Linux v-hub 5.4.0-65-generic #73-Ubuntu SMP Mon Jan 18 17:25:17 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
Linux strongSwan U5.8.2/K5.4.0-65-generic

Do you have any idea what could be the problem?

Thanks, Karel

**Comment by tobiasbrunner** [comment, marked as answer]
Since this has nothing to do with strongSwan, please contact the kernel developers via [netdev mailing list](http://vger.kernel.org/vger-lists.html#netdev).

**Reply by karelvins** [reply]
Many thanks.

Regards,

Karel

**Reply by karelvins** [reply]
https://github.com/torvalds/linux/commit/e98e44562ba2ee994f4fd1e32be7e327edd263ca#diff-6850d259201404fa1f3d1da081f82601579505fe1bc9f30891bedc1abaf82761

**Reply by karelvins** [reply]
I will try to test it with kernel 5.9 or newer where the optimization of xfrm interface lookup is done.
Author of that commit contacted me through netdev mailing list with the same suggestion.

Maybe it shlould be useful to have some note in Strongswan wiki https://wiki.strongswan.org/projects/strongswan/wiki/RouteBasedVPN#XFRM-Interfaces-on-Linux that the performance limitation is there in older kernels.

Regards,

Karel

**Reply by karelvins** [reply]
I tested kernel 5.10 with up to 10 000 XFRM interfaces and can corfirm that it is working very well. Performance degradation is gone.

Karel

**Reply by cyanideangel2000** [reply]
I ran into the same problem and found the commit upstream in kernel 5.9 help alleviate the performance degradation with higher xfrm interfaces. I moved to kernel 5.15 from kernel 5.4 to pull the fix in. Test results are good post this.
